gluten-core/src/main/java/io/glutenproject/memory/memtarget/ThrowOnOomMemoryTarget.java

```diff
@@ -0,0 +1,75 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.glutenproject.memory.memtarget;
+
+import io.glutenproject.memory.MemoryUsageStats;
+
+import org.apache.spark.memory.SparkMemoryUtil;
+
+public class ThrowOnOomMemoryTarget implements MemoryTarget {
+  private final TaskManagedMemoryTarget target;
+
+  public ThrowOnOomMemoryTarget(TaskManagedMemoryTarget target) {
+    this.target = target;
+  }
+
+  @Override
+  public long borrow(long size) {
+    long granted = target.borrow(size);
+    if (granted < size) {
+      // OOM
+      if (granted != 0L) {
+        target.repay(granted);
+      }
+      // Error message
+      StringBuilder errorBuilder = new StringBuilder();
+      errorBuilder.append(
+          String.format(
+              "Not enough spark off-heap execution memory. Acquired: %d, granted: %d. "
+                  + "Try tweaking config option spark.memory.offHeap.size to get larger space "
+                  + "to run this application. %n",
+              size, granted));
+      // Log memory usage
+      target.getTaskMemoryManager().showMemoryUsage();
+      // Dump usages to exception body
+      errorBuilder.append(SparkMemoryUtil.dumpMemoryConsumerStats(target.getTaskMemoryManager()));
+      throw new OutOfMemoryException(errorBuilder.toString());
+    }
+    return granted;
```

Comment:
nit: if (success acquired) { // logic } else { // OOM logic } would more readable.

Reply:
Good suggestion however I'd omit the "else" block with fail-fast in if block. Let me make the change.